### PR TITLE
gh-110850: Add PyTime_t C API

### PR DESCRIPTION
**Pull request by encukou**
This builds on Victor's PR: #112135

Additionally:
- The API of the public functions now allows them to raise exceptions
- The units of `PyTime_t` are now an implementation detail; `PyTime_AsNanoseconds()` is provided to convert to nanoseconds. (Currently, it's a no-op and always succeeds; the API allows reporting errors like overflow in the future.)

* Issue: gh-110850

----
📚 Documentation preview 📚: https://cpython-previews--115215.org.readthedocs.build/

### CODE REVIEW

File: Doc/c-api/time.rst [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+Clock Functions
+---------------
+
+The following functions take a pointer to a :c:expr:`PyTime_t` that they
+set to the value of a particular clock.
+Details of each clock are given in the documentation of the corresponding
+Python function.
+
+The functions return 0 on success, or -1 (with an exception set) on failure.
```
Comment (vstinner) [inline Doc/c-api/time.rst]:
```suggestion
The functions return ``0`` on success, or set an exception and return ``-1`` on failure.
```

File: Doc/c-api/time.rst [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+   Convert a timestamp to a number of seconds as a C :c:expr:`double`.
+
+   The function cannot fail, but note that :c:expr:`double` has limited
+   accuracy for large values.
+
+.. c:function:: int PyTime_AsNanoseconds(PyTime_t t, int64_t *result)
+
+   Convert a timestamp to a number of nanoseconds as a 64-bit signed integer.
+
+   Returns 0 on success, or -1 (with an exception set) on failure.
```
Comment (vstinner) [inline Doc/c-api/time.rst]: This function cannot fail, I don't think that it's worth it add an error case.

Comment (vstinner) [inline Doc/c-api/time.rst]:
```suggestion
   Returns ``0`` on success, or set an exception and return ``-1`` on failure.
```

File: Doc/c-api/time.rst [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+   The unit used is an internal implementation detail; use
+   :c:func:`PyTime_AsNanoseconds` or :c:func:`PyTime_AsSecondsDouble`
+   to convert to a given unit.
+
+   The supported range is at least around [-292.3 years; +292.3 years].
+   Using the Unix epoch (January 1st, 1970) as reference, the supported date
+   range is at least around [1677-09-21; 2262-04-11].
+   The exact limits are exposed as constants.
+
+   .. c:var:: PyTime_t PyTime_MIN
```
Comment (vstinner) [inline Doc/c-api/time.rst]: This variable is declared *inside* PyTime_t, and so Sphinx fails to find it Doc/whatsnew/3.13.rst. Maybe remove the indentation?

File: Include/cpython/pytime.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
+// PyTime_t C API: see Doc/c-api/time.rst for the documentation.
+
+#ifndef Py_LIMITED_API
+#ifndef Py_PYTIME_H
+#define Py_PYTIME_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef PY_INT64_T PyTime_t;
```
Comment (vstinner) [inline Include/cpython/pytime.h]: PY_INT64_T is now always defined as `#define PY_INT64_T int64_t`, so:

```suggestion
typedef int64_t PyTime_t;
```

File: Include/cpython/pytime.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#define PyTime_MIN INT64_MIN
+#define PyTime_MAX INT64_MAX
+
+PyAPI_FUNC(double) PyTime_AsSecondsDouble(PyTime_t t);
```
Comment (vstinner) [inline Include/cpython/pytime.h:14]: nitpick: i suggest to first declare functions "creating" PyTime_t values, and then in second just (just add an empty line) functions to "convert" PyTime_t values.

File: Include/cpython/pytime.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+PyAPI_FUNC(int) PyTime_AsNanoseconds(PyTime_t t, PY_INT64_T *result);
```
Comment (vstinner) [inline Include/cpython/pytime.h]:
```suggestion
PyAPI_FUNC(int) PyTime_AsNanoseconds(PyTime_t t, int64_t *result);
```

File: Include/cpython/pytime.h [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+_PyTime_AsNanoseconds_impl(PyTime_t t, PY_INT64_T *result)
+{
+    // Implementation detail: PyTime_t is already a int64_t in nanoseconds.
+    // This may change. We might e.g. extend PyTime_t to 128 bits, then
+    // this function can start raising OverflowError.
+    *result = t;
+    return 0;
+}
+
+#define PyTime_AsNanoseconds _PyTime_AsNanoseconds_impl
```
Comment (vstinner) [inline Include/cpython/pytime.h]: I'm not sure that it's worth it to add such static inline function which is an optimization.

File: Lib/test/test_capi/test_time.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+CLOCK_RES_NS = int(CLOCK_RES * SEC_TO_NS)
+
+
+class CAPITest(unittest.TestCase):
+    def test_min_max(self):
+        # PyTime_t is just int64_t
+        self.assertEqual(PyTime_MIN, -2**63)
+        self.assertEqual(PyTime_MAX, 2**63 - 1)
+
+    def check_clock(self, c_func, py_func, c_func_ns, py_func_ns):
```
Comment (vstinner) [inline Lib/test/test_capi/test_time.py]: nitpick: it can be moved after test_assecondsdouble().

File: Lib/test/test_capi/test_time.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        # Test PyTime_PerfCounter()
+        self.check_clock(
+            _testcapi.PyTime_PerfCounter, time.perf_counter,
+            _testcapi.PyTime_PerfCounter_ns, time.perf_counter_ns)
+
+    def test_time(self):
+        # Test PyTime_time()
+        self.check_clock(
+            _testcapi.PyTime_Time, time.time,
+            _testcapi.PyTime_Time_ns, time.time_ns)
```
Comment (vstinner) [inline Lib/test/test_capi/test_time.py]: If you want to keep PyTime_AsNanoseconds(), please add a simple test.

File: Modules/_testcapi/time.c [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    }
+    printf("... %d\n", (int)t);
+    return pytime_as_float(t);
+}
+
+static PyObject*
+pytime_as_ns(PyTime_t t)
+{
+    assert(sizeof(long long) >= sizeof(PyTime_t));
+    long long result;
```
Comment (vstinner) [inline Modules/_testcapi/time.c]:
```suggestion
    int64_t result;
```

File: Modules/_testcapi/time.c [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    if (PyTime_AsNanoseconds(t, &result) < 0) {
+        return NULL;
+    }
```
Comment (vstinner) [inline Modules/_testcapi/time.c]: Here you can add:

```
assert(sizeof(result) <= sizeof(long long));
```

File: Modules/_testcapi/time.c [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    printf("... %d\n", (int)t);
+    return pytime_as_ns(t);
+}
+
+
+static PyMethodDef test_methods[] = {
+    {"PyTime_AsSecondsDouble", test_pytime_assecondsdouble, METH_VARARGS},
+    {"PyTime_Monotonic", test_pytime_monotonic, METH_NOARGS},
+    {"PyTime_PerfCounter", test_pytime_perf_counter, METH_NOARGS},
+    {"PyTime_Time", test_pytime_time, METH_NOARGS},
```
Comment (vstinner) [inline Modules/_testcapi/time.c:87]: You can convert nanoseconds to double in Python, what's the purpose of these 3 functions?

The native unit is nanoseconds, I would prefer to only expose test functions using nanoseconds.

File: Python/pytime.c [hunk 13]
```diff
@@ -1405,14 +1396,48 @@ _PyTime_gmtime(time_t t, struct tm *tm)
...
-    _PyTime_t now = _PyTime_GetMonotonicClock();
+    _PyTime_t now;
+    if (PyTime_Monotonic(&now) < 0) {
+        PyErr_Clear();
+        now = 0;
+    }
     return deadline - now;
 }
+
+// Internal wrappers that ignore all excetpions.
```
Comment (vstinner) [inline Python/pytime.c]:
```suggestion
// Internal wrappers that ignore all exceptions.
```

**Review by vstinner — COMMENTED** [review]
Here is a first review.

I suggest to declare that PyTime_t is a number of nanoseconds and not add PyTime_AsNanoseconds().

There are some Sphinx warnings, like: `c:identifier reference target not found: result` in `time.rst`.

**Comment by vstinner** [comment]
We should either:

* Stick to 64-bit and nanoseconds.
* Design the API so that the type can become 128-bit integer and use better resolution than nanoseconds. In this case, I suggest to add 2 functions: PyTime_AsNanoseconds() and PyTime_FromNanoseconds().

File: Doc/c-api/time.rst [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+--------------------
+
+.. c:function:: double PyTime_AsSecondsDouble(PyTime_t t)
+
+   Convert a timestamp to a number of seconds as a C :c:expr:`double`.
+
+   The function cannot fail, but note that :c:expr:`double` has limited
+   accuracy for large values.
+
+.. c:function:: int PyTime_AsNanoseconds(PyTime_t t, int64_t *result)
```
Comment (vstinner) [inline Doc/c-api/time.rst]: The current implementation uses also `PyTime_t` type for `result`, so there is no possible risk of overflow. I suggest to reuse the `PyTime_t` type for that. In this case, the function cannot fail with overflow even if `PyTime_t` size changes later. What do you think?

**Comment by vstinner** [comment]
@encukou: Thanks for working on this PR!

File: Python/pytime.c [hunk 15]
```diff
@@ -1405,14 +1396,48 @@ _PyTime_gmtime(time_t t, struct tm *tm)
...
+// (The exceptions are rare events; we don't care about performance
+// if they're raised.)
+#define NOEXC_WRAPPER(NAME, FUNC)                   \
+    PyTime_t                                        \
+    NAME(void) {                                    \
+        PyTime_t result;                            \
+        if (FUNC(&result) < 0) {                    \
```
Comment (colesbury) [inline Python/pytime.c]: These functions can be called without holding the GIL. I don't this pattern is safe in that case.

**Review by colesbury — COMMENTED** [review]
The fallible APIs and generalization of  `PyTime_t` makes these APIs harder to use without, it seems to me, sufficient benefit:

- API users are likely to start relying on details, like the fact that `PyTime_t` is 64-bit time in nanoseconds. If we ever in the future wanted an API with a different precision or data type, we'd be better off introducing a new API than breaking users.
- The underlying functions shouldn't fail on modern operating systems. Is there any OS we support that doesn't have `CLOCK_MONOTONIC`?

I commented below, but I don't think we can set+clear exceptions in the internal versions. They're used in contexts that don't have active PyThreadStates.

**Comment by colesbury** [comment]
Oh, I see that fallible vs infallible API has already been extensively discussed in the C-API WG. The API doc should probably explicitly state that the functions require holding the GIL.

File: Modules/_testcapi/time.c (hunk 12 above)
Comment (encukou) [reply Modules/_testcapi/time.c:87]: The purpose was to test `PyTime_AsNanoseconds`.

**Comment by vstinner** [comment]
@colesbury:
> Oh, I see that fallible vs infallible API has already been extensively discussed in the C-API WG. The API doc should probably explicitly state that the functions require holding the GIL.

That's the **public** C API. *If* Python needs variants which cannot fail, we can still have some in our *internal* C API.

Right, the rationale for reporting errors was discussed in length in the issue: https://github.com/capi-workgroup/decisions/issues/8.

**Review by vstinner — APPROVED** [review]
LGTM.

For a few seconds (or many nanoseconds ;-)), I tried to imagine how the API and the implementation would look without `PyTime_t` type: using `int64_t` type. It's tempting to make the API as small as possible. IMO the API is *better* with a decided type, even if it's well documented that it's exactly `int64_t`. In terms of typing, it's better, it makes the code easier to follow, to review, etc.

File: Python/pytime.c (hunk 15 above)
Comment (vstinner) [reply Python/pytime.c]: I suggest to remove this macro and keep the 3 existing internal functions which pass `0` to `raise_exc`.

**Comment by encukou** [comment]
Yes: unlike `int64_t`, `PyTime_t` is always in nanoseconds.

**Comment by encukou** [comment]
C API WG is in favour, so I'll merge.

**Comment by encukou** [comment]
Thank you for the code & reviews!

**Comment by vstinner** [comment]
@encukou and me added a bare minimum public C API to read time with a `PyTime_t` type (64-bit signed number: nanoseconds): change commit 879f4546bfbc9c47ef228e7c3d2f126f3d8d64bf.

```c
typedef int64_t PyTime_t;
#define PyTime_MIN INT64_MIN
#define PyTime_MAX INT64_MAX

PyAPI_FUNC(double) PyTime_AsSecondsDouble(PyTime_t t);
PyAPI_FUNC(int) PyTime_Monotonic(PyTime_t *result);
PyAPI_FUNC(int) PyTime_PerfCounter(PyTime_t *result);
PyAPI_FUNC(int) PyTime_Time(PyTime_t *result);
```

If someone wants more functions to be exposed, please open a separated issue.

**Comment by scoder** [comment]
The monotonic and "bare" time functions work as expected, but for the performance counter, I'm getting very different absolute values from the C-API and Python API calls:
```python
    tic1 = time.perf_counter()
    tic2 = ctime.perf_counter()
    tic3 = time.perf_counter()

Failed example:
    tic2 <= tic3  or  (tic2, tic3)
Expected:
    True
Got:
    (1715071333.3882606, 310873.228373332)
```
Both apparently use different reference points, reproducible across runs, as shown by a second run a bit later:
```
Failed example:
    tic2 <= tic3  or  (tic2, tic3)
Expected:
    True
Got:
    (1715071789.5720396, 311329.41215237)
```
Is this expected? Can both be made to use the same absolute values?

**Comment by scoder** [comment]
Sorry, error on my side. Should have talked to my rubber duck first. Sorry for the noise.

**Pull request merged**